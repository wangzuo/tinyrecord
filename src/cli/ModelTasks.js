// @flow
import _ from 'lodash';
import fs from 'fs';
import path from 'path';
import mkdirp from 'mkdirp';
import { plural } from 'pluralize';
import MigrationTasks from './MigrationTasks';

export default class ModelTasks {
  static index() {
    const files = fs.readdirSync(path.join(process.cwd(), 'models'));
    const module = this.createIndex(files);
    const filepath = path.join(process.cwd(), 'models', 'index.js');

    fs.writeFileSync(filepath, module);
  }

  static modelName(name: string) {
    return _.upperFirst(_.camelCase(name));
  }

  static tableName(name: string) {
    return _.toLower(plural(_.snakeCase(name)));
  }

  static create(name: string, attributes) {
    const modelName = this.modelName(name);
    MigrationTasks.create(`Create${plural(modelName)}`, attributes);

    const module = this.createModel(name, attributes);
    const filename = `${modelName}.js`;
    const modelDir = path.join(process.cwd(), 'models');
    const filepath = path.join(modelDir, filename);

    mkdirp.sync(modelDir);
    fs.writeFileSync(filepath, module);

    this.index();
  }

  static createModel(name: string, attributes) {
    const modelName = this.modelName(name);
    const tableName = this.tableName(name);
    const module = `const { Base } = require('tinyrecord');

class ${modelName} extends Base {}
${modelName}.tableName = '${tableName}';

module.exports = ${modelName};
`;

    return module;
  }

  static createIndex(files: Array<string>) {
    return files
      .filter(x => x !== 'index.js')
      .sort()
      .map(file => {
        const module = path.basename(file, '.js');
        return `exports.${module} = require('./${module}');\n`;
      })
      .join('');
  }
}
