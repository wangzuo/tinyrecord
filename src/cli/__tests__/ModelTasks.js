import ModelTasks from '../ModelTasks';

test('modelName', () => {
  expect(ModelTasks.modelName('user')).toBe('User');
  expect(ModelTasks.modelName('User')).toBe('User');
  expect(ModelTasks.modelName('search_history')).toBe('SearchHistory');
  expect(ModelTasks.modelName('SearchHistory')).toBe('SearchHistory');
});

test('tableName', () => {
  expect(ModelTasks.tableName('user')).toBe('users');
  expect(ModelTasks.tableName('User')).toBe('users');
  expect(ModelTasks.tableName('search_history')).toBe('search_histories');
  expect(ModelTasks.tableName('SearchHistory')).toBe('search_histories');
});

test('createModel', () => {
  expect(ModelTasks.createModel('User')).toMatchSnapshot();
  expect(ModelTasks.createModel('SearchHistory')).toMatchSnapshot();
  expect(ModelTasks.createModel('search_history')).toBe(
    ModelTasks.createModel('SearchHistory')
  );
});

test('createIndex', () => {
  expect(
    ModelTasks.createIndex([
      'User.js',
      'Post.js',
      'SearchHistory.js',
      'index.js'
    ])
  ).toMatchSnapshot();
});
