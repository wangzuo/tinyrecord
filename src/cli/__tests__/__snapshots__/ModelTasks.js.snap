// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createIndex 1`] = `
"exports.Post = require('./Post');
exports.SearchHistory = require('./SearchHistory');
exports.User = require('./User');
"
`;

exports[`createModel 1`] = `
"const { Base } = require('tinyrecord');

class User extends Base {}
User.tableName = 'users';

module.exports = User;
"
`;

exports[`createModel 2`] = `
"const { Base } = require('tinyrecord');

class SearchHistory extends Base {}
SearchHistory.tableName = 'search_histories';

module.exports = SearchHistory;
"
`;
